assets/src/edit-story/app/story/actions/useSaveStory.js

```diff
@@ -25,18 +25,37 @@ import { addQueryArgs } from '@wordpress/url';
  */
 import { useAPI } from '../../api';
 import { OutputPage } from '../../../output';
+import { useConfig } from '../../config';
 
 /**
  * Creates AMP HTML markup for saving to DB for rendering in the FE.
  *
- * @param {Object} pages Object of pages.
+ * @param {Object} story Story object.
+ * @param {string} story.featuredMediaUrl Featured media URL.
+ * @param {Array<Object>} pages List of pages.
+ * @param {Object} metadata Metadata.
+ * @param {string} metadata.publisher Publisher name.
+ * @param {string} metadata.publisherLogo Publisher logo.
  * @return {Element} Markup of pages.
  */
-const getStoryMarkupFromPages = (pages) => {
-  const markup = pages.map((page) => {
-    return renderToString(<OutputPage page={page} />);
-  });
-  return markup.join('');
+const getStoryMarkup = (story, pages, metadata) => {
+  // TODO: get different image sizes for featured media.
+
+  return renderToString(
```

Comment:
What about `<html></html>`? My concern is that we still need to put `<script extension="amp-video" src="..../amp-video-0.1.js">` and such. Some times we also need some higher level meta tags in the `head` as well. Is there any other way to merge/combine our `<html><head>` with whatever WP wants to add at serving time? Another issue: this is the only way for us to construct an efficient `<style amp-custom>` as well based on the usage in the page contents.

Reply:
We could construct the whole document in JS and combine it with the server-side parts at serving time, yes. It just means parsing the HTML, merging tags, etc when serving the page. Not very efficient.

But then we should store the extensions and styles in the JSON blob as well (plus the publisher logo, etc). Just so we can always reconstruct the HTML server-side if needed.

Reply:
I know everything always dislikes this, but what about we tried something like:

```
<html>
  ...
  __REPLACE_PUBLISHER_LOGO__
  ...
</html>
```

And then replace such vars (`__REPLACE_PUBLISHER_LOGO__`) with safe HTML escapes w/o parsing the document?

Reply:
Some simple placeholders like this for simply adding some more tags could be done.

The publisher logo is an attribute of `<amp-story>`  as well as the schema microdata, we'd need a way to easily update both server-side if needed.

Perhaps @spacedmonkey has some ideas here?

Reply:
Head tags are still server side rendered ATM, as I can see from the code. I think that this should remain the case, as it useful for head to be changed after first publishing, because of logo / publisher name changes and changes to stylesheets. Hard coding any value in post content, means it would become hard to fix issues after publishing.

Reply:
Correct. The idea here is to generate everything client-side, but have some mechanism server-side to still have that dynamically. The question is just what this should look like. Parsing the whole pre-generated story markup as a DOM document in PHP is computationally expensive, so perhaps we could use some markers as mentioned above to support this more easily.

Reply:
Both parsing as DOM document or regex are both expensive. So some kind of placeholder is really the only way forward.

Reply:
If the tags do not normally conflict, placeholder is definitely a good path for us. Even though we will be generating HTML on JS side, we would still prefer to do bare minimum there: only the things that strictly pertain to AMP.

Reply:
E.g. we could just put:

```
<head>
<!-- __REPLACE_START_HEAD__  -->
...
<!-- __REPLACE_END_HEAD__  -->
</head>
```

Reply:
Let's try that 👍

Reply:
Forgot that one cannot really add HTML comments in a React component without using `dangerouslySetInnerHTML`, but that function needs an actual element.

Reply:
Could we use conveniently named and placed `<meta />` tags, that has no meaning outside of this?

Reply:
(and if anyone suggests XSLT... 😠👊)

Reply:
Hmm why not! `<meta name="web-stories-replace-start-head" />` and similar should work.

Reply:
Could be. But we gotta also be careful: if we put a `<meta>` tag - we must check that AMP Validator allows it.

Reply:
In other words, I'd like the output of JS layer to be fully valid. But enhanceable.

Reply:
> I'd like the output of JS layer to be fully valid

Same here 👍 

Custom `<meta>` tags are allowed. No issues there.